src/Modules/OrchardCore.Commerce/Services/CardPaymentService.cs

```diff
@@ -0,0 +1,157 @@
+using OrchardCore.ContentManagement;
+using OrchardCore.Entities;
+using OrchardCore.Settings;
+using Stripe;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace OrchardCore.Commerce.Services;
+
+public class CardPaymentService : ICardPaymentService
+{
+    private readonly IShoppingCartPersistence _shoppingCartPersistence;
+    private readonly IPriceService _priceService;
+    private readonly PaymentIntentService _paymentIntentService;
+    private readonly IPriceSelectionStrategy _priceSelectionStrategy;
+    private readonly IContentManager _contentManager;
+    private readonly ISiteService _siteService;
+    private readonly IDataProtectionProvider _dataProtectionProvider;
+    private readonly ILogger<CardPaymentService> _logger;
+
+    public CardPaymentService(
+        IShoppingCartPersistence shoppingCartPersistence,
+        IPriceService priceService,
+        IPriceSelectionStrategy priceSelectionStrategy,
+        IContentManager contentManager,
+        ISiteService siteService,
+        IDataProtectionProvider dataProtectionProvider,
+        ILogger<CardPaymentService> logger)
+    {
+        _paymentIntentService = new PaymentIntentService();
+        _shoppingCartPersistence = shoppingCartPersistence;
+        _priceService = priceService;
+        _priceSelectionStrategy = priceSelectionStrategy;
+        _contentManager = contentManager;
+        _siteService = siteService;
+        _dataProtectionProvider = dataProtectionProvider;
+        _logger = logger;
+    }
+
+    public async Task<PaymentIntent> CreatePaymentAsync(ConfirmPaymentRequest request)
+    {
+        var currentShoppingCart = await _shoppingCartPersistence.RetrieveAsync();
+        var totals = await currentShoppingCart.CalculateTotalsAsync(_priceService, _priceSelectionStrategy);
+
+        // Same here as on the checkout page: Later we have to figure out what to do if there are multiple
+        // totals i.e., multiple currencies.
+        var defaultTotal = totals.FirstOrDefault();
+
+        var paymentIntent = new PaymentIntent();
+        var requestOptions =
+            new RequestOptions
+            {
+                ApiKey = (await _siteService.GetSiteSettingsAsync())
+                    .As<StripeApiSettings>()
+                    .SecretKey
+                    .DecryptStripeApiKey(_dataProtectionProvider, _logger),
+            };
+
+        if (request.PaymentMethodId != null)
+        {
+            var paymentIntentOptions = new PaymentIntentCreateOptions
+            {
+                // We need to remove the decimal points and convert the value (decimal) to long.
+                // https://stripe.com/docs/currencies#zero-decimal
+                Amount = long
+                .Parse(
+                    string.Join(
+                        string.Empty,
+                        defaultTotal.ToString().Where(char.IsDigit)),
```

Comment:
You are talking about 3.1415 => 31415, right?

`provided in a currency’s smallest unit`

I don't know if cent is the smallest unit everywhere or not. That's why I made it this way. Because in USD the value never will be 3.1415 but if there is a currency where the cent is not the smallest unit this way it converts it accordingly. Also if it's round like 10 USD the decimal values will be detectable so 10.00$. 

Regarding HUF, TWD, and UGX you are right, those are separate cases. I will correct it to handle those.

Reply:
No, that was only to drive the point home. I'm rather talking about the 3.1 => 31.

The problem with your way is that `.ToString()` removes the trailing zeroes. Here is a fun example: the MSRP of a PS5 is $499.99 so the amount is 49999, but the MSRP of an Xbox Series X is $499.00 so using your algorithm the amount is 499. I prefer a Playstation, but sure I'd buy an Xbox for only $4.99! That's a steal, literally. Also a legal liability nightmare.

Reply:
I see what you are talking about. But I tried this thing you mentioned and when using `.ToString()` on USD for example, it keeps everything. Because when debugging you can see that it's displaying the zero decimal values too: 
![image](https://user-images.githubusercontent.com/80963259/181995800-62b758cf-5be6-475d-9f23-bf322d9712b0.png)

That's why I used this solution because this way if a currency has a lower unit than cents it will display _"X 1.000"_ so converting it will result in a good amount.

`From what I understand the docs say that for most currencies you want to take the cent value`
My only problem with your solution is that there are 135+ supported currencies.  We have 3 groups, the special currencies HUF, TWD, UGX, the zero-decimal currencies (16), and I can only assume that the others are non-zero-decimal currencies and their smallest unit is cent. So we would have to deal with a lot of cases.

Reply:
> But I tried this thing you mentioned and when using .ToString() on USD for example, it keeps everything.

Ah, I misunderstood. I thought this value is `decimal`, but instead it's `Amount`.

> So we would have to deal with a lot of cases.

You have to deal with those cases anyway. `Amount.ToString()` uses `ICurrency.ToString(decimal)` under the hood. This relies on `ICurrency.DecimalPlaces`, which is 2 by default so that assumption already exists in the library. If you want to rely on `Amount.ToString()`, then you have to set the `ICurrency.DecimalPlaces` to 0 for every zero-decimal currency and the special cases. But that's a Stripe-specific implementation detail.
Also I don't think my suggestion is very complicated. You just put all the zero-decimal currency codes in an array, the three special cases in a different array and then handle the 3 different groups in an `if`-`else if`-`else` construct.

Reply:
I did what you wrote here it's working. 
Only one thing, I think for the special cases this is not good:
`(long)Math.Round(default total / 100m) * 100` this is what you wrote.

For the special cases, I think we need to multiply it by 10000 after dividing it by 100 since 10 HUF is 1000. 
So if we have 100 HUF (and we are following the code above) we divide it by 100 (1), then we round it (1), then we multiply it by 100 (100) but that's incorrect since 100 HUF should be 10000 in long.

Reply:
The multiplier is 100, because of [this](https://stripe.com/docs/currencies#:~:text=you%20can%E2%80%99t%20submit%20a%20payout%20for%20the%20full%20balance%2C%20huf%2010.45%2C%20because%20the%20amount%20value%20of%201045%20is%20not%20evenly%20divisible%20by%20100.). Basically they treat the special cases as if they were "normal" currency, but you you can only charge whole numbers. I have no idea why they can't just make them zero-decimal currencies then. Must be for some historical reasons.

Reply:
Good, you are completely correct.